src/codegen/derive_debug.rs

```diff
@@ -0,0 +1,197 @@
+                                }    
+                            });
+
+    match impl_.unwrap().node {
+        ast::ItemKind::Impl(_, _, _, _, _, ref items) => items.clone(),
+        _ => unreachable!(),
+    }
+}
+
+fn gen_field_data_debug_impl(
+    ctx: &BindgenContext,
+    data: &FieldData,
+) -> Option<(String, Vec<TokenTree>)> {
+    if let Some(name) = data.name() {
+        gen_item_debug_impl(ctx, ctx.resolve_item(data.ty()), name)
+    } else {
+        None
+    }
+}
+
+fn gen_bitfield_unit_debug_impl(
+    ctx: &BindgenContext,
+    data: &BitfieldUnit,
+) -> Option<(String, Vec<TokenTree>)> {
+    let mut format_string = String::new();
+    let mut tokens = Vec::new();
+    for (i, bu) in data.bitfields().iter().enumerate() {
+        if i > 0 {
+            format_string.push_str(", ");
+            tokens.push(TokenTree::Token(DUMMY_SP, Token::Comma));
+        }
+        format_string.push_str(&format!("{} : {{:?}}", bu.name()));
+        let name_ident = ctx.rust_ident_raw(bu.name());
+        tokens.extend(quote_tokens!(ctx.ext_cx(), self.$name_ident()));
+    }
+
+    Some((format_string, tokens))
+}
+
+fn gen_item_debug_impl(
+    ctx: &BindgenContext,
+    item: &Item,
+    name: &str,
+) -> Option<(String, Vec<TokenTree>)> {
+    let name_ident = ctx.rust_ident_raw(name);
+
+    let ty = match item.as_type() {
+        Some(ty) => ty,
+        None => {
+            return None;
+        }
+    };
+
+    fn debug_print(
+        ctx: &BindgenContext,
+        name: &str,
+        name_ident: ast::Ident,
+    ) -> Option<(String, Vec<TokenTree>)> {
+        Some((
+            format!("{}: {{:?}}", name),
+            quote_tokens!(ctx.ext_cx(), self.$name_ident),
+        ))
+    }
+
+    match *ty.kind() {
+        // Handle the simple cases.
+        TypeKind::Void |
+        TypeKind::NullPtr |
+        TypeKind::Int(..) |
+        TypeKind::Float(..) |
+        TypeKind::Complex(..) |
+        TypeKind::Function(..) |
+        TypeKind::Enum(..) |
+        TypeKind::Reference(..) |
+        TypeKind::BlockPointer |
+        TypeKind::UnresolvedTypeRef(..) |
+        TypeKind::ObjCInterface(..) |
+        TypeKind::ObjCId |
+        TypeKind::Comp(..) |
+        TypeKind::ObjCSel => debug_print(ctx, name, name_ident),
+
+        TypeKind::TemplateInstantiation(ref inst) => {
+            if inst.is_opaque(ctx, item) {
+                Some((format!("{}: opaque", name), vec![]))
+            } else {
+                debug_print(ctx, name, name_ident)
+            }
+        } 
+
+        // The generic is not required to implement Debug, so we can not debug print that type
+        TypeKind::Named => {
+            Some((format!("{}: Non-debuggable generic", name), vec![]))
+        }
+
+        TypeKind::Array(_, len) => {
+            // Generics are not required to implement Debug
+            if ctx.lookup_item_id_has_type_param_in_array(&item.id()) {
```

Comment:
This should be:

```rust
if item.has_type_param_in_array(ctx) {
    ...
}
```

You may have to add `use ir::analysis::HasTypeParamInArray;` to the top of the file.